File: sentil-core/examples/probabilistic.rs
//! Probabilistic monitoring: lift a noisy sensor into an ensemble and estimate the satisfaction probability with a Wilson interval.
//!
//! Run with `cargo run --example probabilistic`.

use sentil::{LiftingRegistry, Monitor, MonitorConfig, NoiseInteraction, NoiseModel, SmcConfig};

fn main() -> sentil::Result<()> {
    let times: Vec<f64> = (0..20).map(f64::from).collect();
    let mut trace = sentil::Trace::new(times)?;
    trace.add_signal("x", (0..20).map(|i| 0.4 + 0.05 * f64::from(i)).collect::<Vec<_>>())?;

    let mut lifting = LiftingRegistry::new();
    lifting.register("x", NoiseModel::gaussian(0.0, 0.3)?, NoiseInteraction::Additive);

    let config = MonitorConfig::new().smc(SmcConfig {
        samples: 5000,
        ..SmcConfig::default()
    });
    let monitor = Monitor::new("P>=0.9 (always (x > 0))", config)?;
    let result = monitor.check(&trace, &lifting)?;
    println!(
        "probability {:.3}, interval [{:.3}, {:.3}], holds {}",
        result.probability, result.interval.lower, result.interval.upper, result.holds
    );
    Ok(())
}

File: sentil-ffi/examples/probabilistic.c
/* Probabilistic monitoring: lift a noisy sensor and estimate the satisfaction probability. */
#include "sentil.h"
#include <stdio.h>

int main(void) {
    double times[20];
    double xs[20];
    for (int i = 0; i < 20; ++i) {
        times[i] = i;
        xs[i] = 0.4 + 0.05 * i;
    }
    sentil_trace_t *trace = sentil_trace_create(times, 20);
    sentil_trace_add_signal(trace, "x", xs, 20);

    sentil_lifting_registry_t *lifting = sentil_lifting_registry_create();
    sentil_lifting_registry_register(lifting, "x", sentil_noise_gaussian(0.0, 0.3),
                                     SENTIL_NOISE_ADDITIVE);

    sentil_formula_t *phi = sentil_formula_parse("P>=0.9 (always (x > 0))");
    sentil_smc_config_t config = sentil_smc_config_default();
    config.samples = 5000;
    sentil_smc_result_t result;
    if (sentil_formula_check(phi, trace, lifting, &config, &result) != SENTIL_OK) {
        fprintf(stderr, "check error: %s\n", sentil_get_last_error());
        sentil_formula_destroy(phi);
        sentil_lifting_registry_destroy(lifting);
        sentil_trace_destroy(trace);
        return 1;
    }
    printf("probability %.3f, interval [%.3f, %.3f], holds %s\n", result.probability,
           result.interval.lower, result.interval.upper, result.holds ? "true" : "false");

    sentil_formula_destroy(phi);
    sentil_lifting_registry_destroy(lifting);
    sentil_trace_destroy(trace);
    return 0;
}